### PR TITLE
some simple hidpp20 tests

### PR DESCRIPTION
**Pull request by pfps**
Adresses #1097

This PR has some simple hidpp20 tests.  Some are for converting lighting and profile to and from byte strings.  The others are for testing code that uses feature requests.  The functions being tested in the latter tests are the last ones in hidpp20.p

The latter tests use mock but maybe there is a better way to set up these tests.

### CODE REVIEW

**Comment by MattHag** [comment]
The tests can be trimmed down to approximately half the code length via [fixtures](https://docs.pytest.org/en/6.2.x/fixture.html#what-fixtures-are), as they encapsulate the setup code and make it reusable.

**Comment by pfps** [comment]
The function logitech_receiver.hidpp20.feature_request sends a message to the device and returns the response.    Many hidpp20 functions just send messages and decipher the result.

The mocking function (feature_request in the test file) checks to see if the aguments match the first expected ones (that would produce the expected message) and if so returns the reply.   Then it advances to the next expected message.  After the test finishes there is a check to make sure that all the expected calls were used up.  So feature_request needs access to the list of expected arguments and the correct responses, which is only known inside the test function, and has to modify this value each time it is called.

**Comment by MattHag** [comment]
I can help with the first tests, so the code is as simple as possible, easy to maintain and each test case understandable. Let's revise that when the underlaying structure has been improved.

**Comment by pfps** [comment]
Agreed.

**Comment by MattHag** [comment]
I've refactored your hidpp20 tests, removed the global and added a fixture, which can be used as is for functions, that do not return something. Optionally, you can mock responses and assert on the feature_request args and the result.

**Comment by pfps** [comment]
I had to make some changes, as the responses weren't being consumed. 
I added a few tests and cleaned up some others.

**Comment by pfps** [comment]
@MattHag For some reason the tests are failing on ubuntu 3.7.

**Comment by MattHag** [comment]
Yes, that's due to [_.args_ not being available in Python 3.7](https://docs.python.org/3.7/library/unittest.mock.html#unittest.mock.Mock.call_args). It got [introduced with Python 3.8](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.call_args).

**Comment by pfps** [comment]
So switch to 3.8 then?

**Comment by MattHag** [comment]
Here it's easy to fix by index access like `call_args[0]`.

**Comment by MattHag** [comment]
> So switch to 3.8 then?

That's another question, as [Python 3.7 is end-of-life](https://devguide.python.org/versions/). But it is not necessary here. When more complications arise, then 3.7 could be dropped.

File: tests/logitech_receiver/test_profiles.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    assert setting.form == 0x04, "LED setting color read incorrectly"
+    assert setting.intensity == 0x07, "LED setting color read incorrectly"
+    yaml_out = yaml.dump(setting)
+    assert eyaml == re.compile(r"\s+").sub(" ", yaml_out)
+    setting = yaml.safe_load(eyaml)
+    assert setting.to_bytes() == ebytes
+
+
+def test_button_bytes():
+    bbytes = bytes.fromhex("8000FFFF")
```
Comment (MattHag) [inline tests/logitech_receiver/test_profiles.py:38]: If parametrize is not well suitable for this test case, please improve the order a bit. A good approach for ordering code inside a test is

1. Setup and define expected values
2. Run code under test
3. Assert

and to split each of these groups with a new line.

File: tests/logitech_receiver/test_profiles.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    assert setting.period == 0x0050, "LED setting color read incorrectly"
+    assert setting.form == 0x04, "LED setting color read incorrectly"
+    assert setting.intensity == 0x07, "LED setting color read incorrectly"
+    yaml_out = yaml.dump(setting)
+    assert eyaml == re.compile(r"\s+").sub(" ", yaml_out)
+    setting = yaml.safe_load(eyaml)
+    assert setting.to_bytes() == ebytes
+
+
+def test_button_bytes():
```
Comment (MattHag) [inline tests/logitech_receiver/test_profiles.py:37]: If parametrize is not well suitable for this test case, please improve the order a bit. A good approach for ordering code inside a test is

1. Setup and define expected values
2. Run code under test
3. Assert

and to split each of these groups with a new line.

File: tests/logitech_receiver/test_profiles.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    bbytes = bytes.fromhex("8000FFFF")
+    button = hidpp20.Button.from_bytes(bbytes)
```
Comment (MattHag) [inline tests/logitech_receiver/test_profiles.py:39]: New line before and after that line

File: tests/logitech_receiver/test_profiles.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    bbytes = bytes.fromhex("80020454")
+    button = hidpp20.Button.from_bytes(bbytes)
+    assert button.behavior == 0x8, "button profile behaviour read incorrectly"
+    assert button.modifiers == 0x04, "button profile type read incorrectly"
+    bytes_out = button.to_bytes()
+    assert bbytes == bytes_out, "button bytes out not same as bytes in"
+
+
+pbytes = bytes.fromhex(
```
Comment (MattHag) [inline tests/logitech_receiver/test_profiles.py:59]: I slightly favour a function here, that at least ensures, that each test starts with unmodified data.

def get_profile_bytes()

File: tests/logitech_receiver/test_profiles.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    "54004500370000000000000000000000"
+    "00000000000000000000000000000000"
+    "00000000000000000000000000000000"
+    "0A01020300500407000000FFFFFFFFFF"
+    "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"
+    "FFFFFFFFFFFFFFFFFFFFFFFFFF7C81"
+)
+
+
+def test_profile_bytes():
```
Comment (MattHag) [inline tests/logitech_receiver/test_profiles.py:79]: `def test_profile_bytes(profile_bytes):`

This makes it more readable. You immediately know, that this test relies on the fixture.

File: tests/logitech_receiver/test_profiles.py
Comment (MattHag) [inline tests/logitech_receiver/test_profiles.py:1]: If parametrize is not well suitable for this test case, please at least improve the order a bit. A good approach - for every test - is grouping according to
1. Setup and define expected values
2. Run code under test
3. Assert

and to split each of these groups with a new line.

**Review by MattHag — COMMENTED** [review]
Please adopt the second file according to some best practices, already applied in the hidpp20 test.

File: tests/logitech_receiver/test_profiles.py
Comment (pfps) [reply tests/logitech_receiver/test_profiles.py:1]: So then intermediate stages have to be repeated?  That doesn't make sense.  In particular, why separate the first test into two, repeating the first part of the setup?

**Comment by MattHag** [comment]
Looks good, thanks.

**Comment by MattHag** [comment]
This can be merged. Adding tests is a continuous effort and can be done in small iterations.

**Comment by MattHag** [comment]
I found a nice way to setup tests with fixtures, which can be parametrized to avoid duplication and readability:
Factory Fixtures
https://www.inspiredpython.com/article/five-advanced-pytest-fixture-patterns

**Pull request merged**